### ISSUE TITLE
ROCm/gfx1150: memory leak and autotune hang, both in accelerated matmul kernels

### ISSUE THREAD

**Issue by radevgit** [tracel-ai/cubecl#1584]
**Crate:** `cubecl-runtime` / `cubecl-hip` / `cubek-matmul` (via `burn` 0.22.0-pre.3, `burn-rocm`)
**Environment:** AMD Radeon 890M (gfx1150), ROCm 7.2.4, Linux

## Symptom 1: unbounded memory leak

- `bytes_in_use` grows every training iteration, unbounded.
- All growth is one exact allocation size: 1,966,080 bytes. Every other size stays flat.
- `try_reserve` for that size always succeeds; old instances are never marked free.
- Not application code: the identical training code on the CPU (`ndarray`) backend does not leak.
- Checked and cleared by reading the source: gradient accumulators, the optimizer step, and a secondary (non-autodiff) model all handle their tensors correctly.
- The leaking size matches no tensor shape the model actually uses.
- No hidden panic: installed a logger, saw zero warnings while it leaked.
- The allocation call (`do_empty`/`initialize_memory`) is the same generic path every op uses, so a backtrace there can't identify which kernel is responsible. Likely an internal kernel workspace buffer, not a user tensor.

## Symptom 2: hangs / can wedge the GPU during autotune

- First run of a matmul shape the autotuner hasn't seen sometimes hangs completely (no output, no panic).
- Escalated 3 times to a full GPU wedge.
- `CUBECL_AUTOTUNE_LEVEL=minimal` (skips ~19 accelerated CMMA/MMA candidates) fixed it every time.
- Repro below: 3 of 5 runs hung; 2 completed clean.

```rust
use burn::nn::transformer::{TransformerEncoder, TransformerEncoderConfig, TransformerEncoderInput};
use burn::optim::{AdamWConfig, GradientsParams, ModuleOptimizer};
use burn::tensor::{Device, Tensor, TensorData};

fn matmul_step(x: Tensor<4>, w: &Tensor<2>, n_out: usize) -> Tensor<3> {
    let [b, _c, f, t] = x.dims(); // [B, 61, 20, 6]
    let x = x.swap_dims(1, 3).swap_dims(1, 2);
    let x = x.reshape([b * f * t, 61]);
    let x = x.matmul(w.clone()); // K=61, not tile-aligned
    let x = x.reshape([b, f, t, n_out]);
    let x = x.swap_dims(1, 3).swap_dims(1, 2);
    x.reshape([b, t * n_out, f])
}

fn main() {
    let device: Device = Device::rocm(0).autodiff();
    let mut model: TransformerEncoder = TransformerEncoderConfig::new(20, 64, 4, 2).init(&device);
    device.sync().unwrap();
    let w: Tensor<2> = Tensor::from_data(TensorData::new(vec![0.01f32; 61 * 16], [61, 16]), &device);
    let mut optim: ModuleOptimizer = ModuleOptimizer::from(AdamWConfig::new().init());

    for step in 0..1000 {
        let raw = vec![0.1f32; 32 * 61 * 20 * 6];
        let x4 = Tensor::<4>::from_data(TensorData::new(raw, [32, 61, 20, 6]), &device);
        let tokens = matmul_step(x4, &w, 16).require_grad();
        let out = model.forward(TransformerEncoderInput::new(tokens));
        let loss = out.powf_scalar(2.0).sum();
        let grads = GradientsParams::from_grads(loss.backward(), &model);
        model = optim.step(1e-3, model, grads);
        if step % 50 == 0 {
            let u = device.memory_pool_usage().unwrap();
            println!("step={step} allocs={} in_use={:.1}MB", u.number_allocs, u.bytes_in_use as f64 / 1e6);
        }
    }
}
```

## Link between the two

`minimal` fixes both. Best guess: the accelerated matmul kernels are the shared cause on gfx1150.

**Comment by radevgit** [comment]
Withdrawing Symptom 1 (memory leak) — it was my own application bug, not cubecl's. My original claim "does not occur on CPU backends" is false; it reproduces identically on ndarray/CPU.

Root cause: a tokenizer forward ran unconditionally while both its consumers were gated behind a flag that was false in every config. Autodiff still builds a graph for it; since the output never reached the loss, it never reached backward(), so nothing freed it.

.inner() does not free an autodiff graph — measured directly (~57 MB/step without it, ~58 MB/step with it). Only backward(), or never building the graph at all (AutodiffModule::valid() + .inner() inputs), releases it.

Post-fix: bytes_in_use 613→1399MB/500 steps → 39.7MB flat at step 1000. Host RSS on CPU: ~1.7 MB/step → ~0.4 KB/step. Throughput: 5→0.85 it/s decay → flat 2.83 it/s. My CUBECL_AUTOTUNE_LEVEL=minimal claim for this symptom was also wrong — an app-side graph leak isn't something that flag affects.

Symptom 2 (autotune hang) still exists — re-verifying separately.

**Comment by nathanielsimard** [comment]
autotune hang might be bad algorithms running extremely slow because of register spill and a bad configurations. Tbh I think it would be great to just abort running any kernel with register spill when autotuning on runtime where we can detect that.
